src/librustc/hir/mod.rs

```diff
@@ -2369,9 +2369,15 @@ impl CodegenFnAttrs {
         }
     }
 
-    /// True if `#[no_mangle]` or `#[export_name(...)]` is present.
+    /// True if it looks like this symbol needs to be exported, for example:
+    ///
+    /// * `#[no_mangle]` is present
+    /// * `#[export_name(...)]` is present
+    /// * `#[linkage]` is present
     pub fn contains_extern_indicator(&self) -> bool {
-        self.flags.contains(CodegenFnAttrFlags::NO_MANGLE) || self.export_name.is_some()
+        self.flags.contains(CodegenFnAttrFlags::NO_MANGLE) ||
+            self.export_name.is_some() ||
+            self.linkage.is_some()
```

Comment:
This should probably check if `linkage` is `private` or `internal` which are explicitly not exporting.

Reply:
Sure yeah I can update it, but FWIW `#[linkage]` is likely basically not at all working for those sorts of linkages, and in general `#[linkage]` is likely buggier in more ways than this...

Reply:
Yeah, I know. Partitioning doesn't support it well either for example (https://github.com/rust-lang/rust/issues/53929). Still, I guess as long as we don't have a proper plan on how to fix it, we shouldn't introduce additional problems.